### PR TITLE
Services/HTTP: Corrected some error codes and added a few new ones.

### PR DESCRIPTION
**Pull request by Subv**
Use the session data to store information about the HTTP session state.
This is based on reverse engineering of the HTTP module.

~~The AddRequestHeader function is still mostly incorrect, this will be fixed in follow up PRs~~

AddRequestHeader's behavior has been corrected.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4001)

### CODE REVIEW

File: src/core/hle/service/http_c.cpp [hunk 1]
```diff
@@ -11,30 +11,80 @@ namespace HTTP {
...
-    ResultCode(ErrCodes::InvalidContext, ErrorModule::HTTP, ErrorSummary::InvalidState,
+const ResultCode ERROR_STATE_ERROR = // 0xD8A0A066
+    ResultCode(ErrCodes::SessionStateError, ErrorModule::HTTP, ErrorSummary::InvalidState,
                ErrorLevel::Permanent);
 
 void HTTP_C::Initialize(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp(ctx, 0x1, 1, 4);
     const u32 shmem_size = rp.Pop<u32>();
-    rp.PopPID();
+    u32 pid = rp.PopPID();
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: why is this added?

File: src/core/hle/service/http_c.cpp [hunk 2]
```diff
@@ -11,30 +11,80 @@ namespace HTTP {
...
     IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
     // This returns 0xd8a0a046 if no network connection is available.
     // Just assume we are always connected.
     rb.Push(RESULT_SUCCESS);
+}
 
-    LOG_WARNING(Service_HTTP, "(STUBBED) called, shared memory size: {}", shmem_size);
+void HTTP_C::InitializeConnectionSession(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x8, 1, 2);
+    const u32 context_handle = rp.Pop<u32>();
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: const Context::Handle context_handle

File: src/core/hle/service/http_c.cpp [hunk 3]
```diff
@@ -11,30 +11,80 @@ namespace HTTP {
...
+    u32 pid = rp.PopPID();
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: unused pid

File: src/core/hle/service/http_c.cpp [hunk 4]
```diff
@@ -108,19 +210,49 @@ void HTTP_C::AddRequestHeader(Kernel::HLERequestContext& ctx) {
...
+    if (itr->second.state != RequestState::NotStarted) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
+        rb.Push(ResultCode(ErrCodes::InvalidRequestState, ErrorModule::HTTP,
+                           ErrorSummary::InvalidState, ErrorLevel::Permanent));
+        rb.PushMappedBuffer(value_buffer);
         return;
     }
 
-    ASSERT(itr->second.state == RequestState::NotStarted);
     ASSERT(std::find_if(itr->second.headers.begin(), itr->second.headers.end(),
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: Is this assert fine or has the http module an error for that?

File: src/core/hle/service/http_c.cpp (hunk 1 above)
Comment (Subv) [reply src/core/hle/service/http_c.cpp]: we'll likely use it later for verification purposes. For now I wanted to add it to the log but forgot to /actually add it to the log/

File: src/core/hle/service/http_c.cpp (hunk 4 above)
Comment (Subv) [reply src/core/hle/service/http_c.cpp]: I could not find any, so i left the assert in until further investigation is performed

**Comment by B3n30** [comment]
@Subv can you fix that PR so we can merge it soon, I'd like to continue on http_c based on the changes in that PR

File: src/core/hle/service/http_c.cpp [hunk 5]
```diff
@@ -11,30 +11,80 @@ namespace HTTP {
...
         shared_memory->name = "HTTP_C:shared_memory";
     }
 
+    LOG_WARNING(Service_HTTP, "(STUBBED) called, shared memory size: {} pid: {}", shmem_size, pid);
+
+    auto* session_data = GetSessionData(ctx.Session());
+    ASSERT(session_data);
+
+    if (session_data->initialized) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR

File: src/core/hle/service/http_c.cpp [hunk 6]
```diff
@@ -11,30 +11,80 @@ namespace HTTP {
...
+void HTTP_C::InitializeConnectionSession(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x8, 1, 2);
+    const Context::Handle context_handle = rp.Pop<u32>();
+    u32 pid = rp.PopPID();
+
+    auto* session_data = GetSessionData(ctx.Session());
+    ASSERT(session_data);
+
+    if (session_data->initialized) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR

File: src/core/hle/service/http_c.cpp [hunk 7]
```diff
@@ -50,11 +100,45 @@ void HTTP_C::CreateContext(Kernel::HLERequestContext& ctx) {
     LOG_DEBUG(Service_HTTP, "called, url_size={}, url={}, method={}", url_size, url,
               static_cast<u32>(method));
 
+    auto* session_data = GetSessionData(ctx.Session());
+    ASSERT(session_data);
+
+    if (!session_data->initialized) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR

File: src/core/hle/service/http_c.cpp [hunk 8]
```diff
@@ -80,18 +166,34 @@ void HTTP_C::CloseContext(Kernel::HLERequestContext& ctx) {
 
     LOG_WARNING(Service_HTTP, "(STUBBED) called, handle={}", context_handle);
 
+    auto* session_data = GetSessionData(ctx.Session());
+    ASSERT(session_data);
+
+    if (!session_data->initialized) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR

File: src/core/hle/service/http_c.cpp [hunk 9]
```diff
@@ -108,19 +210,49 @@ void HTTP_C::AddRequestHeader(Kernel::HLERequestContext& ctx) {
...
-    // Copy thr value_buffer into a string without the \0 at the end
+    // Copy the value_buffer into a string without the \0 at the end
     std::string value(value_size - 1, '\0');
     value_buffer.Read(&value[0], 0, value_size - 1);
 
+    auto* session_data = GetSessionData(ctx.Session());
+    ASSERT(session_data);
+
+    if (!session_data->initialized) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR

File: src/core/hle/service/http_c.cpp [hunk 10]
```diff
@@ -108,19 +210,49 @@ void HTTP_C::AddRequestHeader(Kernel::HLERequestContext& ctx) {
...
+
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
+        rb.Push(ResultCode(ErrorDescription::NotImplemented, ErrorModule::HTTP,
+                           ErrorSummary::Internal, ErrorLevel::Permanent));
+        rb.PushMappedBuffer(value_buffer);
+        return;
+    }
+
+    if (session_data->current_http_context != context_handle) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR

File: src/core/hle/service/http_c.cpp [hunk 11]
```diff
@@ -108,19 +210,49 @@ void HTTP_C::AddRequestHeader(Kernel::HLERequestContext& ctx) {
...
+
     auto itr = contexts.find(context_handle);
-    if (itr == contexts.end()) {
-        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
-        rb.Push(ERROR_CONTEXT_ERROR);
-        LOG_ERROR(Service_HTTP, "called, context {} not found", context_handle);
+    ASSERT(itr != contexts.end());
+
+    if (itr->second.state != RequestState::NotStarted) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
```
Comment (B3n30) [inline src/core/hle/service/http_c.cpp]: LOG_ERROR
